CAT/strategy/NCAT_strategy.py

```diff
@@ -0,0 +1,665 @@
+        x1 = x1.transpose(1, 2).contiguous() \
+             .view(nbatches, -1, self.h * self.d_k)
+        
+        x2 = x2.transpose(1, 2).contiguous() \
+             .view(nbatches, -1, self.h * self.d_k)
+        return self.linears[-1](x1), self.linears[-1](x2), 
+
+class PositionwiseFeedForward(nn.Module):
+    "Implements FFN equation."
+    def __init__(self, d_model, d_ff, dropout=0.1):
+        super(PositionwiseFeedForward, self).__init__()
+        self.w_1 = nn.Linear(d_model, d_ff)
+        self.w_2 = nn.Linear(d_ff, d_model)
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, x):
+        return self.w_2(self.dropout(F.relu(self.w_1(x))))
+    
+class IRT(nn.Module):
+    def __init__(self, num_students, num_questions, num_dim,config):
+        # num_dim: IRT if num_dim == 1 else MIRT
+        super().__init__()
+        self.config = config
+        self.num_dim = num_dim
+        self.num_students = num_students
+        self.num_questions = num_questions
+        self.theta = nn.Embedding(self.num_students, self.num_dim)
+        self.alpha = nn.Embedding(self.num_questions, self.num_dim)
+        self.beta = nn.Embedding(self.num_questions, 1)
+        for name, param in self.named_parameters():
+            if 'weight' in name:
+                nn.init.xavier_normal_(param)
+
+    def forward(self, student_ids, question_ids):
+        theta = self.theta(student_ids)
+        alpha = self.alpha(question_ids)
+        
+        beta = self.beta(question_ids)
+        pred = (alpha * theta).sum(dim=1, keepdim=True) + beta
+        pred = torch.sigmoid(pred)
+        return pred
+    def adaptest_update(self, adaptest_data: AdapTestDataset):
+        """
+        Update CDM with tested data
+        """
+        lr = self.config['learning_rate']
+        batch_size = self.config['batch_size']
+        epochs = self.config['num_epochs']
+        device = self.config['device']
+        optimizer = torch.optim.Adam(self.theta.parameters(), lr=lr)
+
+        tested_dataset = adaptest_data.get_tested_dataset(last=True)
+        dataloader = torch.utils.data.DataLoader(tested_dataset, batch_size=batch_size, shuffle=True)
+        ls = 0.0
+        for ep in range(1, epochs + 1):
+            
+            loss = 0.0
+            log_steps = 100
+            for cnt, (student_ids, question_ids, _, labels) in enumerate(dataloader):
+                student_ids = student_ids.to(device)
+                question_ids = question_ids.to(device)
+                labels = labels.to(device).float()
+                pred = self.forward(student_ids, question_ids).view(-1)
+                bz_loss = self._loss_function(pred, labels)
+                optimizer.zero_grad()
+                bz_loss.backward(retain_graph=True)
+                optimizer.step()
+                loss += bz_loss.data.float()
+            ls = ls + loss
+        return ls/(epochs)
+    def _loss_function(self, pred, real):
```

Comment:
Please make sure your code is well-formatted, such as keeping a blank line between these functions